### PR TITLE
[CI environment] Retrieve main resource type leveraging api specs

### LINKED ISSUE #3475: Update set-module readme to comply with singular form
Retrieve main resourceType from module relative folder path. Correct resourceType must be in module title

### PR DESCRIPTION
**Pull request by eriqua**
# Description

Closes #3475

Update set-module readme to retrieve main resource type from module folder path leveraging api specs json file
Works with both singular and plural module folder names, hence can be merged to main

Should be merged after #3507 #3510 #3512

| Pipeline |
| - |
| [Storage - StorageAccounts](https://github.com/Azure/ResourceModules/actions/workflows/ms.storage.storageaccounts.yml) |
| [Compute - VirtualMachineScaleSets](https://github.com/Azure/ResourceModules/actions/workflows/ms.compute.virtualmachinescalesets.yml) |

# Type of Change

Please delete options that are not relevant.

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Update to documentation

# Checklist

- [ ] I'm sure there are no other open Pull Requests for the same update/change
- [ ] My corresponding pipelines / checks run clean and green without any errors or warnings
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (readme)
- [ ] I did format my code

### CODE REVIEW

File: modules/insights/action-groups/README.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Action Groups `[Microsoft.Insights/actionGroups]`
```
Comment (AlexanderSehr) [inline modules/insights/action-groups/README.md:1]: That's a bit off an odd one. In the specs file, it is upper case, yet here it is lower. Not sure, if it is worth troubleshooting, but it happened in a few places.

File: utilities/tools/helper/Get-SpecsAlignedResourceName.ps1 [hunk 2]
```diff
@@ -67,25 +67,55 @@ function Get-SpecsAlignedResourceName {
...
     } else {
         $providerNamespace = ($foundProviderNamespaceMatches.Count -eq 1) ? $foundProviderNamespaceMatches : $foundProviderNamespaceMatches[0]
     }
 
     $innerResourceTypes = $specs[$providerNamespace].Keys | Sort-Object
-    $rawResourceType = Get-ReducedWordString -StringToReduce ($reducedResourceIdentifier -replace ('{0}/' -f ($reducedResourceIdentifier.Split('/')[0])), '')
-    $foundResourceTypeMatches = $innerResourceTypes | Where-Object { $_ -like "$rawResourceType*" }
+    $rawResourceTypeReduced = Get-ReducedWordString -StringToReduce $rawResourceType
+    $foundResourceTypeMatches = $innerResourceTypes | Where-Object { $_ -like "$rawResourceTypeReduced*" }
+    # $foundResourceTypeMatches = $innerResourceTypes | Where-Object { $_ -match "$rawResourceTypeReduced(.*?)\\" }
```
Comment (AlexanderSehr) [inline utilities/tools/helper/Get-SpecsAlignedResourceName.ps1:84]:
```suggestion
```

File: modules/insights/action-groups/README.md (hunk 1 above)
Comment (eriqua) [reply modules/insights/action-groups/README.md:1]: I just noticed it is now fixed in main, which is great. It's lowecase in my branch specs file and that's the reason of the changes. I'll merge latest and regen the Microsoft.Insight readmes later.

File: utilities/tools/helper/ConvertTo-ModuleResourceType.ps1 [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    $fullParentResourceType = Get-SpecsAlignedResourceName -ResourceIdentifier $parentResourceIdentifier
+
+    if (-not $childTypeString) {
+        $fullResourceType = $fullParentResourceType
+    } else {
+        $childTypeArray = $childTypeString.Split('/')
+
+        $innerResourceType = $fullParentResourceType
+        foreach ($childType in $childTypeArray) {
+            # Additonal check for child types non existing on their own (e.g. sites/hybridConnectionNamespaces does not exist, sites/hybridConnectionNamespaces/relays does)
```
Comment (eriqua) [inline utilities/tools/helper/ConvertTo-ModuleResourceType.ps1:43]:
```suggestion
            # Additional check for child types non existing on their own (e.g. sites/hybridConnectionNamespaces does not exist, sites/hybridConnectionNamespaces/relays does)
```

File: modules/insights/action-groups/README.md (hunk 1 above)
Comment (AlexanderSehr) [reply modules/insights/action-groups/README.md]: How strange, given that the file is auto-updated in main. Sounds almost like the API changed for like a few days :D

Comment (eriqua) [reply modules/insights/action-groups/README.md]: image

Looks like [it](https://github.com/Azure/ResourceModules/commit/557998bd46373e82282817a199276f532fbaf6b7#diff-b17a07fc484e846ee3d3e558767b0b068c628eb15bd8c3d4338fb79d02011623L17943)
Anyway updated 🔨

File: utilities/tools/helper/ConvertTo-ModuleResourceType.ps1 [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    $provider, $parentType, $childTypeString = $ResourceIdentifier -Split '[\/|\\]', 3
+    $parentResourceIdentifier = $provider, $parentType -join '/'
+
+    $fullParentResourceType = Get-SpecsAlignedResourceName -ResourceIdentifier $parentResourceIdentifier
+
+    if (-not $childTypeString) {
+        $fullResourceType = $fullParentResourceType
+    } else {
+        $childTypeArray = $childTypeString.Split('/')
```
Comment (AlexanderSehr) [inline utilities/tools/helper/ConvertTo-ModuleResourceType.ps1:39]:
```suggestion
        $childTypeArray = $childTypeString -split '\/'
```
For consistency...

File: utilities/tools/helper/Get-SpecsAlignedResourceName.ps1 [hunk 5]
```diff
@@ -67,25 +67,54 @@ function Get-SpecsAlignedResourceName {
 
     $reducedResourceIdentifier = $ResourceIdentifier -replace '-'
 
-    $rawProviderNamespace = $reducedResourceIdentifier.Split('/')[0]
+    $rawProviderNamespace, $rawResourceType = $reducedResourceIdentifier -Split '[\/|\\]', 2
```
Comment (AlexanderSehr) [inline utilities/tools/helper/Get-SpecsAlignedResourceName.ps1:70]:
```suggestion
    $rawProviderNamespace, $rawResourceType = $reducedResourceIdentifier -Split '[\/|\\]', 2 # e.g. 'keyvault' & 'vaults/keys'
```

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**